### PR TITLE
feat: converted setup.{py,cfg} into pyproject.toml

### PR DESCRIPTION
**Pull request by SamuelLarkin**
### PR Goal? 

Migrate `setup.{py,cfg}` to `pyproject.toml` before `pip` forces use to do so.

### Fixes? 

A `pip` deprecation warning.

### Feedback sought? 

Merge approval

### Priority? 

low

### Tests added? 

none

### How to test? 

1. Create a reference environment with `make-everyvoice-env`.
2. Create a new environment with `make-everyvoice-env`.
3. Compare the content of the two environments `diff <(conda env export --name EveryVoice.main_afca68906) <(conda env export --name EveryVoice.pyproject)`.
4. Possibly, examine the logs created during `make-everyvoice-env` but due to some arbitrary order of the logs, this could be tedious.
5. 5. Double check that installing EveryVoice in editable mode still works.

### Confidence? 

fair

### Version change? 

No

### Related PRs? 

None

### CODE REVIEW

**Comment by SamuelLarkin** [comment]
I created a reference conda environment with `main@afca68906` and captured its logs.
I created a new conda environment with the proposed changes.

When comparing the both logs coming out of `make-everyvoice-env`, we see that the **DEPRECATION** warning goes away.

```
DEPRECATION: Legacy editable install of everyvoice==0.2.0a0 from
file:///fs/hestia_Hnrc/ict/sam037/git/EveryVoice (setup.py develop) is
deprecated. pip 25.0 will enforce this behaviour change. A possible
replacement is to add a pyproject.toml or enable --use-pep517, and use
setuptools >= 64. If the resulting installation is not behaving as expected,
try using --config-settings editable_mode=compat. Please consult the
setuptools documentation for more information. Discussion can be found at
https://github.com/pypa/pip/issues/11457
Running setup.py develop for everyvoice
```

When comparing the conda environments,
```sh
vimdiff <(conda env export --name EveryVoice.main_afca68906) <(conda env export --name EveryVoice.pyproject)
```

File: .github/workflows/publish.yaml [hunk 1]
```diff
@@ -21,24 +21,24 @@ jobs:
...
+          token: ${{ secrets.SGILE_PAT }}
+          submodules: recursive
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.10"
+      - name: Install pypa/build
+        run: python3 -m pip install build
+      - name: Build a binary wheel and a source tarball
+        run: python -m build --sdist --wheel
```
Comment (SamuelLarkin) [inline .github/workflows/publish.yaml:36]: Do we want to change to `hatch` or should we wait?
```suggestion
      - name: Install pypa/build
        run: python3 -m pip install hatch
      - name: Build a binary wheel and a source tarball
        run: hatch build
```

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -34,7 +34,7 @@ jobs:
       - name: Install dependencies and package
         run: |
           CUDA_TAG=cpu pip install -r requirements.torch.txt --find-links https://download.pytorch.org/whl/torch_stable.html
-          pip install -r requirements.dev.txt
+          pip install .[dev]
           pip install cython
           pip install -e .
```
Comment (joanise) [inline .github/workflows/test.yml:39]: Can't we replace `pip install .[dev]` and `pip install -e .` by `pip install -e .[dev]`?

Comment (joanise) [reply .github/workflows/test.yml:39]: And if `pip install .[dev]` works without error, we might no longer need `pip install cython`.

File: make-everyvoice-env [hunk 3]
```diff
@@ -158,15 +158,15 @@ r pip install cython
...
+PY_COUNTRY_WORLD=$(grep -o "\(pycountry\|pyworld\)==[0-9.]\+" pyproject.toml)
 if ! r pip install $PY_COUNTRY_WORLD; then
     echo Falling back to installing pycountry and pyworld from conda-forge
     r conda install -y $PY_COUNTRY_WORLD -c conda-forge
 fi
 
 r conda install -y sox -c conda-forge
 r pip install -e .
-r pip install -r requirements.dev.txt
+r pip install .[dev]
```
Comment (joanise) [inline make-everyvoice-env:169]: Might as well turn this into one `-e .[dev]` line.

Comment (joanise) [reply make-everyvoice-env:169]: In fact, doesn't installing `.[dev]` replace the editable install by a non-editable one?

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,163 @@
+[build-system]
+requires = ["hatchling"]
+build-backend = "hatchling.build"
+
+[project]
+name = "EveryVoice"
+description = "Text-to-Speech Synthesis for the Speech Generation for Indigenous Language Education Small Teams Project"
+license = { file = "LICENSE" }
+authors = [
```
Comment (joanise) [inline pyproject.toml:9]: nice, having authors and maintainers like this

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  { name = "Eric Joanis", email = "Eric.Joanis@nrc-cnrc.gc.ca" },
+  { name = "Marc Tessier", email = "Marc.Tessier@nrc-cnrc.gc.ca" },
+  { name = "Mengzhe Geng", email = "Mengzhe.Geng@nrc-cnrc.gc.ca" },
+  { name = "Samuel Larkin", email = "Samuel.Larkin@nrc-cnrc.gc.ca" },
+]
+readme = "README.md"
+requires-python = ">=3.10, <3.12"
+keywords = ["TTS", "CLI"]
+classifiers = [
+  "Development Status :: 5 - Production/Stable",
```
Comment (joanise) [inline pyproject.toml:27]: I don't think we can describe this as Production/Stable... Is there an alpha status?

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+  "Environment :: Console",
+  "Environment :: MacOS X",
+  "Environment :: Other Environment",
+  "Environment :: Win32 (MS Windows)",
```
Comment (joanise) [inline pyproject.toml:31]: I would remove this. While I keep it more or less working on Windows for testing and dev purposes, it's quite broken on Windows in terms of the real functionality. Most importantly, no sox effects. So I would not make this claim.

File: pyproject.toml [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  "Intended Audience :: Education",
+  "Intended Audience :: End Users/Desktop",
+  "Intended Audience :: Other Audience",
+  "Intended Audience :: Science/Research",
+  "License :: OSI Approved :: MIT License",
+  "Operating System :: MacOS :: MacOS X",
+  "Operating System :: MacOS",
+  "Operating System :: Microsoft :: MS-DOS",
+  "Operating System :: Microsoft :: Windows",
+  "Operating System :: Microsoft",
```
Comment (joanise) [inline pyproject.toml:42]: Again, remove all Microsoft and/or Windows reference here.

File: pyproject.toml [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+  "Operating System :: POSIX :: BSD :: FreeBSD",
+  "Operating System :: POSIX :: BSD",
+  "Operating System :: POSIX :: Linux",
+  "Operating System :: POSIX :: SunOS/Solaris",
```
Comment (joanise) [inline pyproject.toml:46]: Never tested on SunOS/Solaris, I would not make this claim.

File: pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+  "Intended Audience :: Other Audience",
+  "Intended Audience :: Science/Research",
+  "License :: OSI Approved :: MIT License",
+  "Operating System :: MacOS :: MacOS X",
+  "Operating System :: MacOS",
+  "Operating System :: Microsoft :: MS-DOS",
+  "Operating System :: Microsoft :: Windows",
+  "Operating System :: Microsoft",
+  "Operating System :: POSIX :: BSD :: FreeBSD",
+  "Operating System :: POSIX :: BSD",
```
Comment (joanise) [inline pyproject.toml:44]: Do we know if this code base works on FreeBSD?

File: pyproject.toml [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+  "Operating System :: POSIX :: Linux",
+  "Operating System :: POSIX :: SunOS/Solaris",
+  "Operating System :: POSIX",
+  "Operating System :: Unix",
```
Comment (joanise) [inline pyproject.toml:48]: ?

File: pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+  "Programming Language :: Python :: 3 :: Only",
+  "Programming Language :: Python :: 3",
+  "Programming Language :: Python :: 3.10",
+  "Programming Language :: Python :: 3.11",
+  "Programming Language :: Python :: 3.12",
```
Comment (joanise) [inline pyproject.toml:53]: Not yet, we're not compatible with Python 3.12 at the moment.

File: pyproject.toml [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+  "Programming Language :: Python",
+  "Programming Language :: Unix Shell",
+  "Topic :: Software Development :: Libraries :: Python Modules",
+  "Topic :: Software Development :: Libraries",
+  "Topic :: Software Development :: User Interfaces",
+  "Topic :: System :: Installation/Setup",
+  "Topic :: System :: Shells",
+  "Topic :: Terminals",
+  "Topic :: Utilities",
+  "Typing :: Typed",
```
Comment (joanise) [inline pyproject.toml:64]: mostly... 😄

File: pyproject.toml [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+]
+dynamic = ["version"]
+
+[project.optional-dependencies]
+# [Specifying GPU version of pytorch for python package in pyproject.toml](https://discuss.pytorch.org/t/specifying-gpu-version-of-pytorch-for-python-package-in-pyproject-toml/209157)
+torch = [
+  # these requirements have to be installed ahead of time in your environment and from a different URL:
+  # CUDA_TAG=cu118 pip install -r requirements.torch.txt --find-links https://download.pytorch.org/whl/torch_stable.html
+  'torch==2.1.0; sys_platform == "darwin"',
+  'torchaudio==2.1.0; sys_platform != "darwin"',
```
Comment (joanise) [inline pyproject.toml:110]: This should be `== darwin`, not `!=`.

File: pyproject.toml [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+  "typer>=0.12.3",
+  "yaspin>=3.1.0",
+]
+dynamic = ["version"]
+
+[project.optional-dependencies]
+# [Specifying GPU version of pytorch for python package in pyproject.toml](https://discuss.pytorch.org/t/specifying-gpu-version-of-pytorch-for-python-package-in-pyproject-toml/209157)
+torch = [
+  # these requirements have to be installed ahead of time in your environment and from a different URL:
+  # CUDA_TAG=cu118 pip install -r requirements.torch.txt --find-links https://download.pytorch.org/whl/torch_stable.html
```
Comment (joanise) [inline pyproject.toml:108]: I guess you didn't find a way to pass a variable in. Darn.

File: pyproject.toml [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+  "isort>=5.10.1",
+  "mypy>=1.8.0",
+  "pre-commit>=3.2.0",
+  "types-pyyaml>=6.0.5",
+  "types-requests>=2.27.11",
+  "types-setuptools>=57.4.9",
+  "types-tabulate==0.9.0",
+  "types-tqdm>=4.64,<5.0",
+  # [Self-referential](https://hatch.pypa.io/dev/config/dependency/#self-referential)
+  #-r requirements.test.txt
```
Comment (joanise) [inline pyproject.toml:125]: I don't feel the need to leave this comment here, it's standard pyproject.toml syntax.

File: pyproject.toml [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+  "everyvoice[test]",
+]
+test = ["jsonschema>=4.17.3", "pep440>=0.1.2"]
+
+[project.scripts]
+everyvoice = "everyvoice.cli:app"
+
+[project.urls]
+Homepage = "https://example.com"
```
Comment (joanise) [inline pyproject.toml:134]: Homepage can be same as repo, for now?

File: pyproject.toml [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+Documentation = "https://docs.everyvoice.ca/latest/"
```
Comment (joanise) [inline pyproject.toml:135]: Why not just `https://docs.everyvoice.ca`?

File: pyproject.toml [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+
+[tool.hatch.version]
+path = "everyvoice/_version.py"
+# pattern = "VERSION = 'b(?P<version>[^']+)'"
+
+[tool.isort]
+ignore = ["E203", "E266", "E501", "W503"]
+max-line-length = 88
+max-complexity = 18
+select = ["B", "C", "E", "F", "W", "T4", "B9"]
```
Comment (joanise) [inline pyproject.toml:157]: These are the flake8 settings.

File: pyproject.toml [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+Changelog = "https://github.com/EveryVoiceTTS/EveryVoice/releases"
+
+[tool.flake8]
+known_first_party = "everyvoice"
+multi_line_output = 3
+include_trailing_comma = true
+force_grid_wrap = 0
+use_parentheses = true
+line_length = 88
+ensure_newline_before_comments = true
```
Comment (joanise) [inline pyproject.toml:147]: These are the isort settings.

**Review by joanise — COMMENTED** [review]
Generally nice. 

And oops, I just noticed you marked this as draft.

I flagged a bunch of things to change form reading the diff, but you probably knew about many of these already, sorry.

I'll also want to actually test.

File: pyproject.toml [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+  { name = "Eric Joanis", email = "Eric.Joanis@nrc-cnrc.gc.ca" },
+  { name = "Marc Tessier", email = "Marc.Tessier@nrc-cnrc.gc.ca" },
+  { name = "Mengzhe Geng", email = "Mengzhe.Geng@nrc-cnrc.gc.ca" },
+  { name = "Samuel Larkin", email = "Samuel.Larkin@nrc-cnrc.gc.ca" },
+]
+maintainers = [
+  { name = "Aidan Pine", email = "Aidan.Pine@nrc-cnrc.gc.ca" },
+  { name = "Eric Joanis", email = "Eric.Joanis@nrc-cnrc.gc.ca" },
+  { name = "Marc Tessier", email = "Marc.Tessier@nrc-cnrc.gc.ca" },
+  { name = "Mengzhe Geng", email = "Mengzhe.Geng@nrc-cnrc.gc.ca" },
```
Comment (joanise) [inline pyproject.toml:20]: Maintainers should be Aidan, Eric and Sam.
Authors should also include Vlad and David, I think.

**Comment by SamuelLarkin** [comment]
# Note:
Comapring `hatch build` with `python -m build --sdist --wheel --outdir dist.build`.

There are no difference
```sh
diff <(cd dist.build/ && sha1sum */**) <(cd dist.hatch/ && sha1sum */**)
```
```sh
delta dist.*
warning: in the working copy of 'dist.build/sdist/everyvoice-0.2.0a0/everyvoice/tests/data/metadata_arpabet.psv', CRLF will be replaced by LF the next time Git touches it
warning: in the working copy of 'dist.hatch/sdist/everyvoice-0.2.0a0/everyvoice/tests/data/metadata_arpabet.psv', CRLF will be replaced by LF the next time Git touches it
warning: in the working copy of 'dist.build/sdist/everyvoice-0.2.0a0/everyvoice/tests/data/metadata_phones.psv', CRLF will be replaced by LF the next time Git touches it
warning: in the working copy of 'dist.hatch/sdist/everyvoice-0.2.0a0/everyvoice/tests/data/metadata_phones.psv', CRLF will be replaced by LF the next time Git touches it
warning: in the working copy of 'dist.build/sdist/everyvoice-0.2.0a0/log.EveryVoice.sl.main_afca68906', CRLF will be replaced by LF the next time Git touches it
warning: in the working copy of 'dist.hatch/sdist/everyvoice-0.2.0a0/log.EveryVoice.sl.main_afca68906', CRLF will be replaced by LF the next time Git touches it
warning: in the working copy of 'dist.build/sdist/everyvoice-0.2.0a0/log.EveryVoice.sl.pyproject', CRLF will be replaced by LF the next time Git touches it
warning: in the working copy of 'dist.hatch/sdist/everyvoice-0.2.0a0/log.EveryVoice.sl.pyproject', CRLF will be replaced by LF the next time Git touches it
warning: in the working copy of 'dist.build/whl/everyvoice/tests/data/metadata_arpabet.psv', CRLF will be replaced by LF the next time Git touches it
warning: in the working copy of 'dist.hatch/whl/everyvoice/tests/data/metadata_arpabet.psv', CRLF will be replaced by LF the next time Git touches it
warning: in the working copy of 'dist.build/whl/everyvoice/tests/data/metadata_phones.psv', CRLF will be replaced by LF the next time Git touches it
warning: in the working copy of 'dist.hatch/whl/everyvoice/tests/data/metadata_phones.psv', CRLF will be replaced by LF the next time Git touches it
```
```sh
difft --skip-unchanged dist.*
```

**Pull request merged**